File: solutions/easy/414.Third_Maximum_Number/414.Third_Maximum_Number.ts
function thirdMax(nums: number[]): number {
  // Sort the array in non-increasing order.
    nums.sort((a, b) => b - a);

    let elemCounted = 1;
    let prevElem = nums[0];

    for (let index = 1; index < nums.length; ++index) {
        // Current element is different from previous.
        if (nums[index] != prevElem) {
            elemCounted += 1;
            prevElem = nums[index];
        }

        // If we have counted 3 numbers then return current number.
        if (elemCounted == 3) {
            return nums[index];
        }
    }

    // We never counted 3 distinct numbers, return largest number.
    return nums[0];
};

File: solutions/easy/414.Third_Maximum_Number/414.Third_Maximum_Number.py
from ast import List


class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        nums.sort(reverse=True)
        elem_counted = 1
        prev_element = nums[0]

        for index in range(1, len(nums)):
            if nums[index] != prev_element:
                elem_counted += 1
                prev_element = nums[index]
            if elem_counted == 3:
                return nums[index]

        return nums[0]
